apps/src/blockly/addons/cdoUtils.ts

```diff
@@ -503,6 +529,44 @@ export function getLevelToolboxBlocks(customCategory: string) {
   }
 }
 
+/**
+ * Creates a map of block types and limits, based on limit attributes found
+ * in the block XML for the current toolbox.
+ * @returns {Map<string, number>} A map of block limits
+ */
+export function createBlockLimitMap() {
+  const parser = new DOMParser();
+  // This method only works for string toolboxes.
+  if (!Blockly.toolboxBlocks || typeof Blockly.toolboxBlocks !== 'string') {
+    return;
+  }
+
+  const xmlDoc = parser.parseFromString(Blockly.toolboxBlocks, 'text/xml');
+  // Define blockLimitMap
+  const blockLimitMap = new Map<string, number>();
+
+  // Select all block elements and convert NodeList to array
+  const blocks = Array.from(xmlDoc.querySelectorAll('block'));
+
+  // Iterate over each block element using forEach
+  blocks.forEach(block => {
+    // Check if the block has a limit attribute
+    const limitAttribute = block.getAttribute('limit');
```

Comment:
I think you can simplify this check to:

```
const limitAttribute = block.getAttribute('limit');
const limit = parseInt(limitAttribute);  // Directly parse the attribute

if (!isNaN(limit)) {
```

`null` and `undefined` should not throw an error in `parseInt` I think and then you can check whether they're numbers.

Reply:
TypeScript actually yells at you if you try `parseInt(limitAttribute)` because `Argument of type 'string | null' is not assignable to parameter of type 'string'.` However, we could probably pass it in as a template string which would coerce `null` to `'null'`.

Reply:
Oh, right! Maybe TypeScript would be happy with
```
const limit = parseInt(block.getAttribute('limit') ?? '');
if (!isNaN(limit)) {
```